bee-tangle/src/urts.rs

```diff
@@ -0,0 +1,197 @@
+        &self.non_lazy_tips
+    }
+
+    pub(crate) async fn insert<B: StorageBackend>(
+        &mut self,
+        tangle: &MsTangle<B>,
+        message_id: MessageId,
+        parents: Vec<MessageId>,
+    ) {
+        if let Score::NonLazy = self.tip_score::<B>(tangle, &message_id).await {
+            self.non_lazy_tips.insert(message_id);
+            self.tips.insert(message_id, TipMetadata::new());
+            for parent in parents.iter() {
+                self.add_child(*parent, message_id);
+                self.check_retention_rules_for_parent(parent);
+            }
+        }
+    }
+
+    fn add_child(&mut self, parent: MessageId, child: MessageId) {
+        match self.tips.entry(parent) {
+            Entry::Occupied(mut entry) => {
+                let metadata = entry.get_mut();
+                metadata.children.insert(child);
+                if metadata.time_first_child == None {
+                    metadata.time_first_child = Some(Instant::now());
+                }
+            }
+            Entry::Vacant(entry) => {
+                let mut metadata = TipMetadata::new();
+                metadata.children.insert(child);
+                metadata.time_first_child = Some(Instant::now());
+                entry.insert(metadata);
+            }
+        }
+    }
+
+    fn check_retention_rules_for_parent(&mut self, parent: &MessageId) {
+        if self.non_lazy_tips.len() > MAX_LIMIT_NON_LAZY as usize
+            || self.tips.get(parent).unwrap().children.len() as u8 > MAX_NUM_CHILDREN
+            || self
+                .tips
+                .get(parent)
+                .unwrap()
+                .time_first_child
+                .unwrap()
+                .elapsed()
+                .as_secs() as u8
+                > MAX_AGE_SECONDS_AFTER_FIRST_CHILD
+        {
+            self.tips.remove(parent);
+            self.non_lazy_tips.remove(parent);
+        }
+    }
+
+    pub(crate) async fn update_scores<B: StorageBackend>(&mut self, tangle: &MsTangle<B>) {
+        let mut to_remove = Vec::new();
+
+        for tip in self.tips.keys() {
+            match self.tip_score::<B>(tangle, &tip).await {
+                Score::SemiLazy | Score::Lazy => {
+                    to_remove.push(*tip);
+                }
+                _ => continue,
+            }
+        }
+
+        for tip in to_remove {
+            self.tips.remove(&tip);
+            self.non_lazy_tips.remove(&tip);
+        }
+
+        debug!("Non-lazy tips {}", self.non_lazy_tips.len());
+    }
+
+    async fn tip_score<B: StorageBackend>(&self, tangle: &MsTangle<B>, message_id: &MessageId) -> Score {
+        // in case the tip was pruned by the node, consider tip as lazy
+        if !tangle.contains(message_id).await {
+            return Score::Lazy;
+        }
+
+        let smi = *tangle.get_solid_milestone_index();
+        let omrsi = *tangle.omrsi(&message_id).await.unwrap().index();
```

Comment:
Maybe document these `unwrap`s.

Reply:
Also these unwraps are safe since we only add `solid` tips to the tip pool
All of the solid tips do have omris/ymrsi values set. So unwrapping should be safe.
We need this as a comment here

Reply:
@rufsam Do you think you could come up with a comment-worth justification for them? It seems like that explanation is relying on some very non-local reasoning

Reply:
Maybe comment something like:

"The tip pool only works with solid tips. Therefore, all tips added to the pool can be considered to solid. The solid flag will be set together with `omrsi` and `ymrsi` values. Therefore, when a message is solid, `omrsi` and `ymrsi` values are available. Therefore, unwrapping here is safe."